Complete the exercises below For **Assignment #13**.

Load the `ISLR2` and the `tidymodels` packages.

In [1]:
library("ISLR2")
library("tidymodels")

── Attaching packages ────────────────────────────────────── tidymodels 1.2.0 ──

✔ broom        1.0.7     ✔ recipes      1.1.0
✔ dials        1.3.0     ✔ rsample      1.2.1
✔ dplyr        1.1.4     ✔ tibble       3.2.1
✔ ggplot2      3.5.1     ✔ tidyr        1.3.1
✔ infer        1.0.7     ✔ tune         1.2.1
✔ modeldata    1.4.0     ✔ workflows    1.1.4
✔ parsnip      1.2.1     ✔ workflowsets 1.1.0
✔ purrr        1.0.2     ✔ yardstick    1.3.1

── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
✖ purrr::discard() masks scales::discard()
✖ dplyr::filter()  masks stats::filter()
✖ dplyr::lag()     masks stats::lag()
✖ recipes::step()  masks stats::step()
• Use suppressPackageStartupMessages() to eliminate package startup messages



In this assignment we will use the `Default` dataset which includes the default status for credit card customers (`default` variable) in addition to each customer's:

1. credit card balance (`balance` variable),
1. student status (`student` variable), and,
1. income (`income` variable).

In [2]:
Default |> head()

default student balance   income   
1 No      No       729.5265 44361.625
2 No      Yes      817.1804 12106.135
3 No      No      1073.5492 31767.139
4 No      No       529.2506 35704.494
5 No      No       785.6559 38463.496
6 No      Yes      919.5885  7491.559

We will be modeling `default` with the customer features.

Before we begin let's count how many customers fall into each `default` category.

In [3]:
Default |> count(default)

default n   
1 No      9667
2 Yes      333

The data is quite imbalanced. This will be important to keep in mind when we evaluate the performance of our model later.

Run the code below to create and training data from `Default`. We will use the "test" dataset at the end to get a final evaluation of our best model's accuracy.

In [4]:
# For reproducibility
set.seed(123)

Default_split = initial_split(Default, prop = 0.90, strata = default)

Default_train = training(Default_split)
Default_test = testing(Default_split)

Create a logistic regression model called `mod`. Set the engine to `glm` and the mode to `classification`.

In [5]:
# Define model
mod =
  logistic_reg() |>
  set_engine("glm") |>
  set_mode("classification")

Our data is imbalanced. As such, a naive model that *always* predicts a customer to **not default** would be correct quite often. Let's start by calculating the "accuracy" of a naive model. This will be the baseline accuracy by which we evaluate other models.

In [6]:
# This code calculates the accuracy of a model that always predicts default to be "No"

Default_train |>
    mutate(.pred_naive = factor('No', levels = c('No', 'Yes'))) |>
    accuracy(truth = default, .pred_naive)

.metric  .estimator .estimate
1 accuracy binary     0.9666667

Let's use k-fold cross validation to evaluate the performance of a model where the outcome is `default` and the predictors are `income` and `balance`.

To start, use `vfold_cv` to generate 10 validation folds (i.e. set the `v` variable to 10). Set the `strata` argument to `default` so we preserve the distribution of `default` values in each fold.

Creat your folds below and use `glimpse` to look at the output table. Call your output folds tables "folds".

In [7]:
# For reproducibility
set.seed(123)

# Define folds
folds =
  vfold_cv(
  data=Default_train,
  v=10,
  strata="default"
)

# Display folds
glimpse(folds)

Rows: 10
Columns: 2
$ splits <list> [<vfold_split[8100 x 900 x 9000 x 4]>], [<vfold_split[8100 x 9…
$ id     <chr> "Fold01", "Fold02", "Fold03", "Fold04", "Fold05", "Fold06", "Fo…


The code below fits a model to each of your 10 folds. `collect_metrics` finds the average of evaluation metrics for each of your ten models.

In [8]:
mod |>
  fit_resamples(default ~ income + balance, folds) |>
  collect_metrics()

.metric     .estimator mean       n  std_err      .config             
1 accuracy    binary     0.97400000 10 0.0011135775 Preprocessor1_Model1
2 brier_class binary     0.02117421 10 0.0008563604 Preprocessor1_Model1
3 roc_auc     binary     0.94947381 10 0.0039990016 Preprocessor1_Model1

❓How does the model accuracy compare to the naive model from above?

The mean accuracy for the logistic regression model (≈97.4%) is superior to the naive model (≈96.7%).

Complete the cell below to evaluate a model also includes the `student` variable as as predictor.
1. use `default ~ income + balance + student` as the formula,
2. encode your `student` variable with `step_dummy`, and,
3. don't forget to `prep` your recipe!

In [9]:
rec =
  recipe(default ~ income + balance + student, data=Default_train) |>
  step_dummy(student)

mod |>
  fit_resamples(rec, folds) |>
  collect_metrics()

.metric     .estimator mean       n  std_err      .config             
1 accuracy    binary     0.97366667 10 0.0012289378 Preprocessor1_Model1
2 brier_class binary     0.02108349 10 0.0008581557 Preprocessor1_Model1
3 roc_auc     binary     0.94968406 10 0.0039625258 Preprocessor1_Model1

❓Does it appear that the model that includes `student` improves upon the first model with only `income` and `balance` as predictors?

No, including `student` does not appear to improve upon oour first model with only `income` and `balance` as predictors. The new metric scores either lower or are roughly unchanged.

Finally, estimate the accuracy of an `default ~ income + balance` model on the test data, `Default_test`.

❓Does our model outperform a naive model?

In [10]:
# Fit the model and augment with test data
augmented =
  mod |>
  fit(default ~ income + balance, data = Default_train) |>
  augment(Default_test)

# Calculate and display accuracy
accuracy = mean(augmented$.pred_class == augmented$default)

accuracy

[1] 0.97

Our logistic model does outperform the naive model based on accuracy.